scripts/2_match_households_and_individuals.py

```diff
@@ -687,184 +686,205 @@ def get_interim_path(
+            for key, value in matches_hh_level.items()
+            if value
+            and not pd.isna(
+                np.random.choice(value)
+            )  # Ensure the value list is not empty and the selected match is not NaN
+        }
 
-    logger.info("Categorical matching: Random sampling complete")
+        # Multiple matches in case we want to try stochastic runs
 
-    # Save results
-    logger.info("Categorical matching: Saving results")
-    # random sample
-    with open(
-        get_interim_path("matches_hh_level_categorical_random_sample.pkl"), "wb"
-    ) as f:
-        pkl.dump(matches_hh_level_sample, f)
+        # Same logic as above, but repeat it multiple times and store each result as a separate dictionary in a list
+        matches_hh_level_sample_list = [
+            {
+                key: np.random.choice(value)
+                for key, value in matches_hh_level.items()
+                if value and not pd.isna(np.random.choice(value))
+            }
+            for i in range(25)  # Repeat the process 25 times
+        ]
 
-    # multiple random samples
-    with open(
-        get_interim_path("matches_hh_level_categorical_random_sample_multiple.pkl"),
-        "wb",
-    ) as f:
-        pkl.dump(matches_hh_level_sample_list, f)
+        logger.info("Categorical matching: Random sampling complete")
+
+        # Save results
+        logger.info("Categorical matching: Saving results")
+
+        # matching results
+        with open(get_interim_path("matches_hh_level_categorical.pkl"), "wb") as f:
+            pkl.dump(matches_hh_level, f)
+
+        # random sample
+        with open(
+            get_interim_path("matches_hh_level_categorical_random_sample.pkl"), "wb"
+        ) as f:
+            pkl.dump(matches_hh_level_sample, f)
+
+        # multiple random samples
+        with open(
+            get_interim_path("matches_hh_level_categorical_random_sample_multiple.pkl"),
+            "wb",
+        ) as f:
+            pkl.dump(matches_hh_level_sample_list, f)
+    else:
+        logger.info("Categorical matching: loading matched households")
+        # Load matching result
+        with open(
+            get_interim_path("matches_hh_level_categorical_random_sample.pkl"), "rb"
+        ) as f:
+            matches_hh_level_sample = pkl.load(f)
+
+        # multiple random samples
+        with open(
+            get_interim_path("matches_hh_level_categorical_random_sample_multiple.pkl"),
+            "rb",
+        ) as f:
+            matches_hh_level_sample_list = pkl.load(f)
+
+    # TODO: check if this:
+    #   - column is required and possibly update other scripts to add this column in-memory since it is large
+    #   - or can use the single sample hh for the new column
+    # For now, updated to use the sample dictionary
+    ## add matches_hh_level as a column in spc_edited
+    spc_edited["nts_hh_id"] = spc_edited["hid"].map(matches_hh_level_sample)
```

Comment:
In #61, this is updated to the sample value instead of the full list. The full list is now output as a separate file above on L852-L853. If this change sounds ok to you @Hussein-Mahfouz, I'll go ahead and remove the TODO here.

Reply:
That sounds good to me